### ISSUE TITLE
Product import - url rewrites are incorrectly generated for multiple storeviews

### ISSUE THREAD

**Issue by hostep** [magento/magento2#8396]
Hi

Yes, I know, there are a lot of issues already open around this, but nothing is happening from Magento's side to try to fix this (as far as I'm seeing).
So here is another case which demonstrates this. Hopefully Magento will try looking into this.

We have this issue happening during a product import which runs programmatically, but I wanted to create a very simple case using the product import csv feature in the Magento backend as to demonstrate this.

It has been driving me crazy, I spend over 20 hours debugging the url rewrites functionality and find it way too complicated and fixing a thing in one place, breaks something in another, and I'm done trying to find a solution on my own, so hopefully the Magento devs will finally take a look at this.
That would be very very very much appreciated!

### Preconditions
- Magento CE 2.1.3 (also tested on develop branch, rev: a2cf9b8a071370fe462ade50fc00fc7e2cce8634)
- PHP 7.0.15

### Steps to reproduce
1. Setup a clean Magento CE 2.1.3 instance
2. Make sure you have 2 storeviews, one with code `en` and another with code `nl`
3. Create 1 category by hand, make sure you enter the name and url_key as follows, also never create redirects when you change the storeview specific url_key:

| Scope | Name             | Url Key          |
| ----- | ---------------- | ---------------- |
| admin | category-admin   | category-admin   |
| en    | category-english | category-english |
| nl    | category-dutch   | category-dutch   |

4. Go to the Import feature in the backend of Magento and set these values (leave the rest as the defaults):
  - Entity Types: Products
  - Import Behavior: Add/Update

5. Upload this csv file
```csv
product_websites,store_view_code,attribute_set_code,product_type,categories,sku,price,name,url_key
base,,Default,simple,Default Category/category-english,product,123,product,product
,en,Default,simple,,product,,product-english,product-english
,nl,Default,simple,,product,,product-dutch,product-dutch
```

### Expected result
These url rewrites should exist in the `url_rewrite` table:

| request_path                          | target_path                          | store_id |
| ------------------------------------- | ------------------------------------ | -------- |
| category-english.html                 | catalog/category/view/id/3           | 1        |
| category-dutch.html                   | catalog/category/view/id/3           | 2        |
| product-english.html                  | catalog/product/view/id/1            | 1        |
| product-dutch.html                    | catalog/product/view/id/1            | 2        |
| category-english/product-english.html | catalog/product/view/id/1/category/3 | 1        |
| **category-dutch/product-dutch.html**     | catalog/product/view/id/1/category/3 | 2        |

### Actual result
These incorrect url rewrites exist in the `url_rewrite` table:

| request_path                          | target_path                          | store_id |
| ------------------------------------- | ------------------------------------ | -------- |
| category-english.html                 | catalog/category/view/id/3           | 1        |
| category-dutch.html                   | catalog/category/view/id/3           | 2        |
| product-english.html                  | catalog/product/view/id/1            | 1        |
| product-dutch.html                    | catalog/product/view/id/1            | 2        |
| category-english/product-english.html | catalog/product/view/id/1/category/3 | 1        |
| **category-english/product-dutch.html**     | catalog/product/view/id/1/category/3 | 2        |

### More remarks
If we do this import programmatically, we actually see the value `category-admin` being used in the url rewrites, not `category-english` but that's because we import the products using the `category-admin` value.

During my debugging efforts, I ran into various caches in certain classes which caches category data (like the url_key), but this cache is never store dependent, so it only keeps the value of one scope in cache, which is most likely what causes this problem I guess. 

Please ignore the fact that in this import csv, I use `Default Category/category-english`. Apparently that is another bug, because using `Default Category/category-admin` creates a new category during the import, but using `Default Category/category-english` uses the one which already exists.
I'm not looking for a solution for this bug, so you can create a new issue for this one if you want, I only want a solution for the incorrect url rewrites.

Thanks!

**Comment by naydav** [comment]
We've created internal ticket MAGETWO-64295 to address this issue.
Thanks

**Comment by TommyKolkman** [comment]
@naydav Is there any chance we could see progress on this? This is really blocking my projects at the moment, since all my efforts to fix this fail, just like @hostep.

**Comment by YevSent** [comment]
Merged into develop branch. I've created a ticket to backport the fix to 2.1.x version (`MAGETWO-64295`).

**Comment by magento-engcom-team** [comment]
@hostep, thank you for your report.
The issue is already fixed in 2.2.0
